fix(otel logging): Include ILogger CategoryName in logger output

johncrim opened the pull request:
Previously the `CategoryName` field from the OpenTelemetry `LogRecord` was omitted, which was not consistent with normal ILogger behavior, or with Application Insights classic behavior.

fixes: https://github.com/Azure/azure-sdk-for-net/issues/44746

johncrim committed:
fix(telemetry transmission): Include ILogger CategoryName in logger output

Previously this field was omitted, which was not consistent with normal ILogger behavior, or with Application Insights classic behavior.

fixes: #44746

johncrim committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net

johncrim committed:
improve(otel exporter tests): Verify that logger CategoryName is in Exception log entries

cijothomas commented on sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:
```diff
@@ -124,6 +124,12 @@ internal static List<TelemetryItem> OtelToAzureMonitorLogs(Batch<LogRecord> batc
+            var categoryName = logRecord.CategoryName;
+            if (!string.IsNullOrEmpty(categoryName))
+            {
+                properties.Add("CategoryName", categoryName.Truncate(SchemaConstants.KVP_MaxValueLength)!);
```
It does match the Classic ApplicationInsights behavior : https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/LOGGING/src/ILogger/ApplicationInsightsLogger.cs#L159 , but I am not sure if this is the plan going forward! CategoryName is like the instrumentation-scope from OTel, so it should be stored in a more neutral way than "CategoryName", which is very ILogger/.NET specific

johncrim replied:
Well, it seems to me to be a big deal to just drop a key component of the log data just because it's not in the Otel documentation. What would the rationale be for not logging it?

Also `CategoryName` is the property on the `LogRecord` type, so it's at least been implemented in the OpenTelemetry main repo.

I'm not trying to be hostile about this, I just think it's important to include.

cijothomas replied:
> What would the rationale be for not logging it?

None. It should be logged. But there is no top-level field in TraceTelemetry to put it, so customProperty is the only option. Then it brings the question - what should be the name of the "key".? There is no clear answer as of today. "CategoryName" was used in classic app insights, but that cannot used as-is for OTel. Because "LoggerCategory" is the InstrumentationScope in OTel, so it must be stored consistently across all languages (node/python/java etc.)

> Also CategoryName is the property on the LogRecord type, so it's at least been implemented in the OpenTelemetry main repo.

OTel .NET specially treats `ILogger` and hence it is stored in LogRecord directly. When exporting to OTel's own OTLP Exporter, they are exported as InstrumentationScope only, *not as attributes*.

johncrim replied:
Thanks for the reply @cijothomas. I looked a little and didn't find a field name in the otel specs/docs for "InstrumentationScope"; and it looks like it's a set of key-value pairs, which if put in a single field would make querying difficult/inefficient. Maybe something like `scope.name` would work, and would align with the conventions.

cijothomas replied:
> didn't find a field name in the otel specs/docs for "InstrumentationScope";

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-instrumentationscope

https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L71

OTel data model has well known way of storing the instrumentationscope. AzureMonitor don't have, and that is why we need to address it (consistently across langauges).

johncrim replied:
Thanks - I had seen the first doc you referenced (which doesn't provide a schema), but not the protobuf schema (which does). There are 2 parts here that seem relevant:

https://github.com/open-telemetry/opentelemetry-proto/blob/40b3c1b746767cbc13c2e39da3eaf1a23e54ffdd/opentelemetry/proto/logs/v1/logs.proto#L66C1-L81C2

Shows that the formalized InstrumentationScope is within a parent object that also contains all the `LogRecords` - so it's not part of each `LogRecord` in this schema. This schema seems possibly useful for in-proc batching or sending LogRecords over the wire, but not for storing flattened records, so I'll assume it would be flattened into the AI `TraceTelemetry`.

https://github.com/open-telemetry/opentelemetry-proto/blob/40b3c1b746767cbc13c2e39da3eaf1a23e54ffdd/opentelemetry/proto/common/v1/common.proto#L69C1-L81C2

shows the schema for InstrumentationScope .

Since I don't think more radical changes to App Insights storage design are likely, I think using a property `scope.name` is the most practical option. Or another option is to stick with `CategoryName` until a top-level InstrumentationScope map is added (if that seems likely), to avoid multiple schema changes.

cijothomas replied:
> Since I don't think more radical changes to App Insights storage design are likely, 

I won't rule that out. For OTel Resource, AppInsights has made a totally new design! I don't know if the owners will chose to do similar for InstrumentationScope or not!

> using a property scope.name is the most practical option

Yes, but it is not really my call to make that decision!

cijothomas replied:
@vishweshbankwar updated that it is decided by the Product team to use the same key as before for now. Resolving this conversation.

TimothyMothra reviewed (changes requested):
Hi @johncrim, Thanks for your contribution! I think this is a reasonable approach. 

I'm blocking this PR so we can take some time to discuss this.
We've been mostly focused on mapping OTel concepts to Azure Monitor. "CategoryName" has been missed because it's an ILogger/.NET concept. 
This is on the agenda to discuss in our next group meeting and I hope to have an update next week.

cc: @vishweshbankwar

johncrim commented:
I should note that this PR doesn't include any changes to the real-time monitoring client. Once an approach is settled on, it seems likely that the same change should be added there.

johncrim commented:
Sorry for the hiccup here - I was just trying to rename my branch (the source of this PR) and accidentally closed the PR. Re-opening.

cijothomas reviewed (approved):
This is aligned with the classic application insights sdk behavior : https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/LOGGING/src/ILogger/ApplicationInsightsLogger.cs#L159

I understand there are future plans to better support scope name (across signals), but this is a reasonable balance for now, given how important `ILogger` category is for `ILogger` users.

cijothomas commented:
@johncrim Could you resolve the merge conflicts? 

I am proposing to merge this and include for the imminent release train, but will let @TimothyMothra make the final decision.

johncrim commented:
@cijothomas  - I haven't looked at the real-time monitoring client much, but I think the same change should be made there. I hadn't done that due to uncertainty around naming.

Taking a look at that work right now...

johncrim committed:
Merge latest changes from main

johncrim commented:
Hi @cijothomas - I've resolved the merge conflicts in the PR.

I __have not__ made any changes to LiveMetrics. I would be happy to - doesn't look hard - but I would like confirmation that the livemetrics trace document and exception document should have a `CategoryName` property to match the normal telemetry.

johncrim commented:
I found that `CategoryName` is included in livemetrics using AppInsights classic, so I'm going to add that to this PR.

johncrim committed:
improve(live metrics): Include LogRecord.CategoryName in live metrics

It's included in live metrics using app insights classic, so should be included here too.

vishweshbankwar reviewed (approved):
LGTM

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

rajkumar-rangaraj reviewed (approved):
@johncrim Could you please update the changelog for both exporter and distro?

johncrim committed:
docs(changelog): Update changelog with "add CategoryName" info

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.
